openedx/features/discussion_provider_piazza/views.py

```diff
@@ -0,0 +1,170 @@
+"""
+Piazza Discussion Provider views.
+"""
+from collections import namedtuple
+from typing import Dict
+from urllib.parse import quote
+
+from django.contrib.auth.models import User
+from django.contrib.sites.shortcuts import get_current_site
+from django.http import HttpRequest
+from lti_consumer.lti_1p1.contrib.django import lti_embed
+from opaque_keys.edx.keys import CourseKey
+from web_fragments.fragment import Fragment
+
+from lms.djangoapps.courseware.access import get_user_role
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.core.djangoapps.discussions.api.config import get_course_discussion_config
+from openedx.core.djangoapps.discussions.api.data import CourseDiscussionConfigData
+from openedx.core.djangoapps.plugin_api.views import EdxFragmentView
+from openedx.core.djangolib.markup import HTML
+from student.models import anonymous_id_for_user
+
+
+LtiCredentials = namedtuple('LtiCredentials', ('oauth_key', 'oauth_secret'))
+
+
+class LtiCourseLaunchMixin:
+    """
+    Mixin that encapsulates all LTI-related functionality from the View
+    """
+
+    LTI_LAUNCH_URL = None
+    ROLE_MAP = {
+        'student': u'Student',
+        'staff': u'Administrator',
+        'instructor': u'Instructor',
+    }
+    DEFAULT_ROLE = u'Student'
+
+    @staticmethod
+    def get_additional_lti_parameters(course_key: CourseKey, request: HttpRequest) -> Dict[str, str]:
+        return {}
+
+    @staticmethod
+    def get_user_id(user: User, course_key: CourseKey):
+        return quote(anonymous_id_for_user(user, course_key))
+
+    def get_lti_roles(self, user: User, course_key: CourseKey) -> str:
+        return self.ROLE_MAP.get(
+            get_user_role(user, course_key),
+            self.DEFAULT_ROLE,
+        )
+
+    @staticmethod
+    def get_context_id(course_key: CourseKey) -> str:
+        return quote(str(course_key))
+
+    @staticmethod
+    def get_resource_link_id(course_key: CourseKey, request: HttpRequest) -> str:
+        site = get_current_site(request)
+        return quote('{}-{}'.format(
+            site.domain,
+            str(course_key.make_usage_key('course', course_key.run)),
+        ))
```

Comment:
This needs to be different for every placement, so perhaps instead of using the course key and converting it to a usage key this can take a usage key directly, and another method can build that.